src/Channels/UnownedBufferChannel.cs

```diff
@@ -0,0 +1,352 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Channels
+{
+    /// <summary>
+    /// Channel which works in buffers which it does not own, as opposed to using a <see cref="IBufferPool"/>. Designed
+    /// to allow Streams to be easily adapted to <see cref="IReadableChannel"/> via <see cref="System.IO.Stream.CopyToAsync(System.IO.Stream)"/>
+    /// </summary>
+    public class UnownedBufferChannel : IReadableChannel, IReadableBufferAwaiter
+    {
+        private static readonly Action _awaitableIsCompleted = () => { };
+        private static readonly Action _awaitableIsNotCompleted = () => { };
+
+        private static Task _completedTask = Task.FromResult(0);
+
+        private Action _awaitableState;
+
+        private BufferSegment _head;
+        private BufferSegment _tail;
+
+        private bool _consuming;
+
+        // REVIEW: This object might be getting a little big :)
+        private readonly TaskCompletionSource<object> _readingTcs = new TaskCompletionSource<object>();
+        private readonly TaskCompletionSource<object> _writingTcs = new TaskCompletionSource<object>();
+        private readonly TaskCompletionSource<object> _startingReadingTcs = new TaskCompletionSource<object>();
+
+        private Gate _readWaiting = new Gate();
+
+        /// <summary>
+        /// Constructs a new instance of <see cref="UnownedBufferChannel" />
+        /// </summary>
+        public UnownedBufferChannel()
+        {
+            _awaitableState = _awaitableIsNotCompleted;
+        }
+
+        /// <summary>
+        /// A <see cref="Task"/> that completes when the consumer starts consuming the <see cref="IReadableChannel"/>.
+        /// </summary>
+        public Task ReadingStarted => _startingReadingTcs.Task;
```

Comment:
This is not a property on IReadableChannel (at least any longer). For the purposes that ReadingStarted is used for in the PR, I think you can just await _readWaiting instead. At the point ReadingStarted is awaited, the buffer is already passed to WriteAsync, so your not saving much in the way of resources by not using _readWaiting instead.